File: Course 3/course3_2020_07_27/week4_paths_in_graphs2/1_minimum_flight_cost/dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>

using std::vector;
using std::queue;
using std::pair;
using std::priority_queue;

typedef pair<int, long long> map;
struct comparator {
  bool operator() (
    map const& a,
    map const& b)
    const
  {
    return a.second > b.second;
  }
};

long long distance(vector<vector<int> > &adj, vector<vector<int> > &cost, int s, int t) {
  //write your code her
  int n = adj.size();
  priority_queue<map, vector<map>, comparator > pq;
  long long* dist = new long long[n];
  bool* visited = new bool[n];
  for (int i = 0; i < n; i++) {
    dist[i] = -1;
    visited[i] = false;
  }
  dist[s] = 0;
  pq.push(std::make_pair(s, 0));
  while(!pq.empty()) {
    map m = pq.top();
    pq.pop();
    int ind = m.first;
    long long d = m.second;
    if (visited[ind]) continue;

    if (ind == t) return d;
    else visited[ind] = true;
    for(int i = 0; i < adj[ind].size(); i++) {
      int temp_ind = adj[ind][i];
      long long c = cost[ind][i];
      if (!visited[temp_ind]) {
        if (dist[temp_ind] == -1 || dist[temp_ind] > d + c) {
          dist[temp_ind] = d + c;
          pq.push(std::make_pair(temp_ind, dist[temp_ind]));
        }
      }
    }
  }
  return -1;
}

int main() {
  int n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  vector<vector<int> > cost(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y, w;
    std::cin >> x >> y >> w;
    adj[x - 1].push_back(y - 1);
    cost[x - 1].push_back(w);
  }
  int s, t;
  std::cin >> s >> t;
  s--, t--;
  std::cout << distance(adj, cost, s, t);
}

File: Course 3/course3_2020_07_27/week4_paths_in_graphs2/1_minimum_flight_cost/Dijkstra.java
import java.util.*;

public class Dijkstra {
    static class Pair<T, R> {
        T first;
        R second;
        public Pair(T x, R y) {
            first = x;
            second = y;
        }
    }
    static Comparator<Pair<Integer, Long>> comparator = new Comparator<Pair<Integer,Long>>() {
        @Override
        public int compare(Pair<Integer, Long> a, Pair<Integer, Long> b){
            if (a.second > b.second) return 1;
            else if (a.second < b.second) return -1;
            else return 0;
        }
    };
    private static long distance(ArrayList<Integer>[] adj, ArrayList<Integer>[] cost, int s, int t) {
        int n = adj.length;
        
        PriorityQueue<Pair<Integer, Long>> pQueue = new PriorityQueue<>(1, comparator);
        long[] dist = new long[n];
        boolean[] visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            dist[i] = -1;
            visited[i] = false;
        }
        dist[s] = 0;
        pQueue.add(new Pair<Integer,Long>(s, dist[s]));
        while(!pQueue.isEmpty()) {
            Pair<Integer, Long> e = pQueue.poll();
            int ind = e.first;
            long d = e.second;
            if (visited[ind]) continue;

            if (ind == t) return dist[ind];
            else visited[ind] = true;
            for (int i = 0; i < adj[ind].size(); i++) {
                int temp_ind = adj[ind].get(i);
                long c = cost[ind].get(i);
                if (!visited[temp_ind]) {
                    if (dist[temp_ind] == -1 || dist[temp_ind] > d + c) {
                        dist[temp_ind] = d + c;
                        pQueue.add(new Pair<Integer,Long>(temp_ind, dist[temp_ind]));
                    }
                } 
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = new ArrayList[n];
        ArrayList<Integer>[] cost = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
            cost[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y, w;
            x = scanner.nextInt();
            y = scanner.nextInt();
            w = scanner.nextInt();
            adj[x - 1].add(y - 1);
            cost[x - 1].add(w);
        }
        int x = scanner.nextInt() - 1;
        int y = scanner.nextInt() - 1;
        System.out.println(distance(adj, cost, x, y));
        scanner.close();
    }
}

